File: 07-selecao-enumeracoes-e-estruturas/exemplos/ponto_final.cpp
#include "bscpp.hpp"
#include <string>

using namespace std;

// Análise
//
// Colocar um ponto final em um texto se ela não terminar com ponto final.
//
// Definição de tipos de dados
//
// O texto é representada por uma string.



// Produz s se s é vazia ou termina com um ponto final, caso contrário
// produz s seguido de um ponto final.
//
// Requer que s não termine com espaço.
string coloca_ponto_se_necessario(string s)
{
    string r;
    // s é vazia ou o último caractere é "."
    if (s == "" || s.substr(s.length() - 1, 1) == ".") {
        r = s;
    } else {
        r = s + ".";
    }
    return r;
}

examples
{
    // Coloca o ponto
    // s.substr(s.length() - 1, 1) == "." é false
    check_expect(coloca_ponto_se_necessario("casa"), "casa.");
    // s.substr(s.length() - 1, 1) == "." é false
    check_expect(coloca_ponto_se_necessario("eu tambem"), "eu tambem.");
    // Não coloca o ponto
    // s.substr(s.length() - 1, 1) == "." é true
    check_expect(coloca_ponto_se_necessario("casa."), "casa.");
    // s == "" é true
    check_expect(coloca_ponto_se_necessario(""), "");
}

int main()
{
    run_tests();
}

File: 04-selecao-enumeracoes-e-estruturas/exemplos/ponto_final.py
# Análise
#
# Colocar um ponto final em um texto caso ele não termine com um.

# Definição de tipos de dados
#
# O texto é representado por uma string

def ponto_final(texto: str) -> str:
    '''
    Produz *texto* se *texto* é vazio ou termina com ponto final, caso
    contrário, produz *texto* concatenado com '.'.

    Exemplos

    Adiciona o ponto

    >>> ponto_final('Sim, eu gostaria')
    'Sim, eu gostaria.'

    Não adiciona o ponto
    >>> ponto_final('')
    ''
    >>> ponto_final('Talvez.')
    'Talvez.'
    '''
    if texto == '' or texto[-1] == '.':
        com_ponto = texto
    else:
        com_ponto = texto + '.'
    return com_ponto
